# Contents

> importing libraries and data

> creat subset

> aggregations

> create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value

> Spending habits of the three types of customers

> create a spend flag

> create a frequency flag

# Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import Data

In [2]:
path = r'/Users/vanessalimsingh/Instacart Basket Analysis 07-05-2024'

In [3]:
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_version2.pkl'))

# Subset of first 1,000,000 rows

In [4]:
df=ords_prods_merge[:1000000]

In [5]:
ords_prods_merge.shape

(32404859, 22)

In [6]:
df.shape

(1000000, 22)

In [7]:
#group by product name
df.groupby('product_name')

In [8]:
#performing an aggregation
# Split the data into groups based on “department_id.”
#Apply the agg() function to each group to obtain the mean values for the “order_number” column.
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

In [9]:
#The table should list all the possible “department_id” values from the current dataframe together 
#with their corresponding “order_number” means—in other words, the average number of orders per user 
#for each department ID.

In [10]:
#another way of aggregating. 
#replace with a command that uses mean function
#this achieves the same result
df.groupby('department_id')['order_number'].mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: order_number, dtype: float64

In [11]:
#multiple aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                14.800024   1  99
2                17.091743   1  98
3                17.913544   1  99
4                17.893092   1  99
5                15.214270   1  99
6                15.382135   1  99
7                17.694027   1  99
8                16.458105   1  91
9                15.957363   1  99
10               20.091818   1  99
11               16.482026   1  99
12               15.615061   1  99
13               16.484023   1  99
14               17.524632   1  99
15               15.691875   1  99
16               18.014071   1  99
17               16.150593   1  99
18               19.602850   1  99
19               17.631340   1  99
20               17.138607   1  99
21               21.956893   1  97

In [12]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/1g/ch7w9_k56pj78tjf9gzlt8zm0000gn/T/ipykernel_16374/1681540670.py:1: FutureWarning: The provided callable <function max at 0x111777920> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [13]:
ords_prods_merge=ords_prods_merge.drop(columns=['Unnamed: 0.2','Unnamed: 0.1', 'Unnamed: 0_x', 'Unnamed: 0_y','prior_busiest_day'])

In [14]:
ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0            both  Mid-range product  Regularly busy        Average orders   
1            both  Mid-range product  Regularly busy        Average orders   
2            both  Low-range product  Regularly busy        Average orders   
3            both  Low-range product  Regularly busy        Average orders   
4            both  Low-range product  Regularly busy        Average orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10

In [15]:
ords_prods_merge.head(100)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2539329        1             1          2                  8   
2    2539329        1             1          2                  8   
3    2539329        1             1          2                  8   
4    2539329        1             1          2                  8   
..       ...      ...           ...        ...                ...   
95    738281        2             4          2                 10   
96   1673511        2             5          3                 11   
97   1673511        2             5          3                 11   
98   1673511        2             5          3                 11   
99   1673511        2             5          3                 11   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                      NaN       14084                  2          0   
2                      NaN       12427                  3          0   
3                      NaN       26088                  4          0   
4                      NaN       26405                  5          0   
..                     ...         ...                ...        ...   
95                     8.0       21150                 13          0   
96                     8.0       47144                  1          0   
97                     8.0        5322                  2          0   
98                     8.0       17224                  3          0   
99                     8.0       38656                  4          0   

                                         product_name  aisle_id  \
0                                                Soda        77   
1             Organic Unsweetened Vanilla Almond Milk        91   
2                                 Original Beef Jerky        23   
3                          Aged White Cheddar Popcorn        23   
4                    XL Pick-A-Size Paper Towel Rolls        54   
..                                                ...       ...   
95                            Fire Grilled Steak Bowl        38   
96     Unsweetened Original Almond Breeze Almond Milk        91   
97  Gluten Free Dark Chocolate Chunk Chewy with a ...         3   
98                   Oats & Honey Gluten Free Granola         3   
99  Gluten Free Peanut Butter Dark Chocolate Chewy...         3   

    department_id  prices merge_indicator    price_range_loc    busiest_days  \
0               7     9.0            both  Mid-range product  Regularly busy   
1              16    12.5            both  Mid-range product  Regularly busy   
2              19     4.4            both  Low-range product  Regularly busy   
3              19     4.7            both  Low-range product  Regularly busy   
4              17     1.0            both  Low-range product  Regularly busy   
..            ...     ...             ...                ...             ...   
95              1     5.9            both  Mid-range product  Regularly busy   
96             16    14.0            both  Mid-range product  Regularly busy   
97             19     2.9            both  Low-range product  Regularly busy   
98             19     1.6            both  Low-range product  Regularly busy   
99             19     3.4            both  Low-range product  Regularly busy   

   busiest_period_of_day  max_order  
0         Average orders         10  
1         Average orders         10  
2         Average orders         10  
3         Average orders         10  
4         Average orders         10  
..                   ...        ...  
95           Most orders         14  
96           Most orders         14  
97           Most orders         14  
98           Most orders         14  
99           Most orders         14  

[100 rows x 18 columns]

In [16]:
#remove head row limit from pandas' default
pd.options.display.max_rows = None

In [17]:
ords_prods_merge.head(100)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2539329        1             1          2                  8   
2    2539329        1             1          2                  8   
3    2539329        1             1          2                  8   
4    2539329        1             1          2                  8   
5    2398795        1             2          3                  7   
6    2398795        1             2          3                  7   
7    2398795        1             2          3                  7   
8    2398795        1             2          3                  7   
9    2398795        1             2          3                  7   
10   2398795        1             2          3                  7   
11    473747        1             3          3                 12   
12    473747        1             3          3                 12   
13    473747        1             3          3                 12   
14    473747        1             3          3                 12   
15    473747        1             3          3                 12   
16   2254736        1             4          4                  7   
17   2254736        1             4          4                  7   
18   2254736        1             4          4                  7   
19   2254736        1             4          4                  7   
20   2254736        1             4          4                  7   
21    431534        1             5          4                 15   
22    431534        1             5          4                 15   
23    431534        1             5          4                 15   
24    431534        1             5          4                 15   
25    431534        1             5          4                 15   
26    431534        1             5          4                 15   
27    431534        1             5          4                 15   
28    431534        1             5          4                 15   
29   3367565        1             6          2                  7   
30   3367565        1             6          2                  7   
31   3367565        1             6          2                  7   
32   3367565        1             6          2                  7   
33    550135        1             7          1                  9   
34    550135        1             7          1                  9   
35    550135        1             7          1                  9   
36    550135        1             7          1                  9   
37    550135        1             7          1                  9   
38   3108588        1             8          1                 14   
39   3108588        1             8          1                 14   
40   3108588        1             8          1                 14   
41   3108588        1             8          1                 14   
42   3108588        1             8          1                 14   
43   3108588        1             8          1                 14   
44   2295261        1             9          1                 16   
45   2295261        1             9          1                 16   
46   2295261        1             9          1                 16   
47   2295261        1             9          1                 16   
48   2295261        1             9          1                 16   
49   2295261        1             9          1                 16   
50   2550362        1            10          4                  8   
51   2550362        1            10          4                  8   
52   2550362        1            10          4                  8   
53   2550362        1            10          4                  8   
54   2550362        1            10          4                  8   
55   2550362        1            10          4                  8   
56   2550362        1            10          4                  8   
57

In [18]:
#create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [19]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'


In [20]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [21]:
# print the frequency of your new “loyalty_flag”
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [22]:
#write a head() function that returned the first 60 rows of only the 
#“user_id,” “loyalty_flag,” and “order_number” columns
ords_prods_merge[['user_id','loyalty_flag','order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
5         1      New customer             2
6         1      New customer             2
7         1      New customer             2
8         1      New customer             2
9         1      New customer             2
10        1      New customer             2
11        1      New customer             3
12        1      New customer             3
13        1      New customer             3
14        1      New customer             3
15        1      New customer             3
16        1      New customer             4
17        1      New customer             4
18        1      New customer             4
19        1      New customer             4
20        1      New customer             4
21        1      New customer             5
22        1      New customer             5
23        1      New customer             5
24        1      New customer             5
25        1      New customer             5
26        1      New customer             5
27        1      New customer             5
28        1      New customer             5
29        1      New customer             6
30        1      New customer             6
31        1      New customer             6
32        1      New customer             6
33        1      New customer             7
34        1      New customer             7
35        1      New customer             7
36        1      New customer             7
37        1      New customer             7
38        1      New customer             8
39        1      New customer             8
40        1      New customer             8
41        1      New customer             8
42        1      New customer             8
43        1      New customer             8
44        1      New customer             9
45        1      New customer             9
46        1      New customer             9
47        1      New customer             9
48        1      New customer             9
49        1      New customer             9
50        1      New customer            10
51        1      New customer            10
52        1      New customer            10
53        1      New customer            10
54        1      New customer            10
55        1      New customer            10
56        1      New customer            10
57        1      New customer            10
58        1      New customer            10
59        2  Regular customer             1

In [23]:
ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0            both  Mid-range product  Regularly busy        Average orders   
1            both  Mid-range product  Regularly busy        Average orders   
2            both  Low-range product  Regularly busy        Average orders   
3            both  Low-range product  Regularly busy        Average orders   
4            both  Low-range product  Regularly busy        Average orders   

   max_order  loyalty_flag  
0         10  New customer  
1         10  New customer  
2         10  New customer  
3         10  New customer  
4         10  New customer

# Task step 2

In [24]:
#aggregated mean of the “order_number” column grouped by “department_id”

In [25]:
ords_prods_merge.shape

(32404859, 19)

In [26]:
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [27]:
#aggregation on the subset
df_subset=df.groupby('department_id').agg({'order_number': ['mean']})

In [28]:
df_subset

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

# 3. The aggregation done on the entire dataset has a higher range.

> the aggregation done on the entire dataset has a range of 15.45 to 22.90

> the aggregation on the subset has a range of 14.8 to 21.96


# 4. Loyalty Flag

In [29]:
#create a loyalty flag for existing customers using the transform() and loc() functions.

In [30]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/1g/ch7w9_k56pj78tjf9gzlt8zm0000gn/T/ipykernel_16374/1681540670.py:1: FutureWarning: The provided callable <function max at 0x111777920> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [31]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [32]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [33]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [34]:
ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0            both  Mid-range product  Regularly busy        Average orders   
1            both  Mid-range product  Regularly busy        Average orders   
2            both  Low-range product  Regularly busy        Average orders   
3            both  Low-range product  Regularly busy        Average orders   
4            both  Low-range product  Regularly busy        Average orders   

   max_order  loyalty_flag  
0         10  New customer  
1         10  New customer  
2         10  New customer  
3         10  New customer  
4         10  New customer

# 5.  Spending habits of the three types of customers

In [35]:
#Use the loyalty flag you created and check the basic statistics of the product prices 
#for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 
#What you’re trying to determine is whether the prices of products purchased 
#by loyal customers differ from those purchased by regular or new customers

In [36]:
# group by new customer ; loyal customer ; regular customer. 
# find the mean of prices for each group

In [37]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

# The prices of products purchased by loyal customers are lower than new customers and regular customers. 

> New customers have the highest price of products purchased, followed by regular customers.

# 6.  Target different types of spenders in their marketing campaigns

In [38]:
#This can be achieved by looking at the prices of the items people are buying. 
#Create a spending flag for each user based on the average price across all their orders using the criteria:


In [39]:
#If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
#If the mean of the prices of products purchased by a user is higher than or equal to 10; 'High Spender'

In [40]:
#create column that shows mean 'prices' of 'user_id'

In [41]:
ords_prods_merge2=ords_prods_merge

In [42]:
ords_prods_merge2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0            both  Mid-range product  Regularly busy        Average orders   
1            both  Mid-range product  Regularly busy        Average orders   
2            both  Low-range product  Regularly busy        Average orders   
3            both  Low-range product  Regularly busy        Average orders   
4            both  Low-range product  Regularly busy        Average orders   

   max_order  loyalty_flag  
0         10  New customer  
1         10  New customer  
2         10  New customer  
3         10  New customer  
4         10  New customer

In [43]:
ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0            both  Mid-range product  Regularly busy        Average orders   
1            both  Mid-range product  Regularly busy        Average orders   
2            both  Low-range product  Regularly busy        Average orders   
3            both  Low-range product  Regularly busy        Average orders   
4            both  Low-range product  Regularly busy        Average orders   

   max_order  loyalty_flag  
0         10  New customer  
1         10  New customer  
2         10  New customer  
3         10  New customer  
4         10  New customer

In [44]:
ords_prods_merge2['ave_prices'] = ords_prods_merge2.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/1g/ch7w9_k56pj78tjf9gzlt8zm0000gn/T/ipykernel_16374/4188781185.py:1: FutureWarning: The provided callable <function mean at 0x111784360> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge2['ave_prices'] = ords_prods_merge2.groupby(['user_id'])['prices'].transform(np.mean)


In [47]:
ords_prods_merge2.head(10)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   
5   2398795        1             2          3                  7   
6   2398795        1             2          3                  7   
7   2398795        1             2          3                  7   
8   2398795        1             2          3                  7   
9   2398795        1             2          3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   
5                    15.0         196                  1          1   
6                    15.0       10258                  2          0   
7                    15.0       12427                  3          1   
8                    15.0       13176                  4          0   
9                    15.0       26088                  5          1   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                     Soda        77              7     9.0   
6                               Pistachios       117             19     3.0   
7                      Original Beef Jerky        23             19     4.4   
8                   Bag of Organic Bananas        24              4    10.3   
9               Aged White Cheddar Popcorn        23             19     4.7   

  merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0            both  Mid-range product  Regularly busy        Average orders   
1            both  Mid-range product  Regularly busy        Average orders   
2            both  Low-range product  Regularly busy        Average orders   
3            both  Low-range product  Regularly busy        Average orders   
4            both  Low-range product  Regularly busy        Average orders   
5            both  Mid-range product  Regularly busy        Average orders   
6            both  Low-range product  Regularly busy        Average orders   
7            both  Low-range product  Regularly busy        Average orders   
8            both  Mid-range product  Regularly busy        Average orders   
9            both  Low-range product  Regularly busy        Average orders   

   max_order  loyalty_flag  ave_prices  
0         10  New customer    6.367797  
1         10  New customer    6.367797  
2         10  New customer    6.367797  
3         10  New customer    6.367797  
4         10  New customer    6.367797  
5         10  New customer    6.367797  
6         10  New customer    6.367797  
7         10  New customer    6.367797  
8         10  New customer    6.367797  
9         10  New customer    6.367797

In [48]:
ords_prods_merge2.loc[ords_prods_merge2['ave_prices'] < 10, 'spend_flag'] = 'low_spender'
ords_prods_merge2.loc[ords_prods_merge2['ave_prices']>= 10, 'spend_flag'] = 'high_spender'

In [54]:
ords_prods_merge2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  department_id  prices  \
0                                     Soda  ...              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk  ...             16    12.5   
2                      Original Beef Jerky  ...             19     4.4   
3               Aged White Cheddar Popcorn  ...             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls  ...             17     1.0   

   merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0             both  Mid-range product  Regularly busy        Average orders   
1             both  Mid-range product  Regularly busy        Average orders   
2             both  Low-range product  Regularly busy        Average orders   
3             both  Low-range product  Regularly busy        Average orders   
4             both  Low-range product  Regularly busy        Average orders   

  max_order  loyalty_flag ave_prices   spend_flag  
0        10  New customer   6.367797  low_spender  
1        10  New customer   6.367797  low_spender  
2        10  New customer   6.367797  low_spender  
3        10  New customer   6.367797  low_spender  
4        10  New customer   6.367797  low_spender  

[5 rows x 21 columns]

In [55]:
ords_prods_merge = ords_prods_merge2

In [56]:
ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  department_id  prices  \
0                                     Soda  ...              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk  ...             16    12.5   
2                      Original Beef Jerky  ...             19     4.4   
3               Aged White Cheddar Popcorn  ...             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls  ...             17     1.0   

   merge_indicator    price_range_loc    busiest_days busiest_period_of_day  \
0             both  Mid-range product  Regularly busy        Average orders   
1             both  Mid-range product  Regularly busy        Average orders   
2             both  Low-range product  Regularly busy        Average orders   
3             both  Low-range product  Regularly busy        Average orders   
4             both  Low-range product  Regularly busy        Average orders   

  max_order  loyalty_flag ave_prices   spend_flag  
0        10  New customer   6.367797  low_spender  
1        10  New customer   6.367797  low_spender  
2        10  New customer   6.367797  low_spender  
3        10  New customer   6.367797  low_spender  
4        10  New customer   6.367797  low_spender  

[5 rows x 21 columns]

# 7. frequent vs non-frequent customers

In [58]:
#Create an order frequency flag that marks the regularity of a user’s ordering behavior according 
#to the median in the “days_since_prior_order” column

In [59]:
#If the median of “days_since_prior_order” is higher than 20, 
#then the customer should be labeled a “Non-frequent customer.”

In [61]:
#If the median is higher than 10 and lower than or equal to 20, 
#then the customer should be labeled a “Regular customer.”

In [62]:
#If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [64]:
#new column added median_days_since_prior_order
ords_prods_merge2['median_days_since_prior_order'] = ords_prods_merge2.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/1g/ch7w9_k56pj78tjf9gzlt8zm0000gn/T/ipykernel_16374/1433722626.py:1: FutureWarning: The provided callable <function median at 0x1118c8400> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge2['median_days_since_prior_order'] = ords_prods_merge2.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [65]:
ords_prods_merge2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  prices  merge_indicator  \
0                                     Soda  ...     9.0             both   
1  Organic Unsweetened Vanilla Almond Milk  ...    12.5             both   
2                      Original Beef Jerky  ...     4.4             both   
3               Aged White Cheddar Popcorn  ...     4.7             both   
4         XL Pick-A-Size Paper Towel Rolls  ...     1.0             both   

     price_range_loc    busiest_days busiest_period_of_day max_order  \
0  Mid-range product  Regularly busy        Average orders        10   
1  Mid-range product  Regularly busy        Average orders        10   
2  Low-range product  Regularly busy        Average orders        10   
3  Low-range product  Regularly busy        Average orders        10   
4  Low-range product  Regularly busy        Average orders        10   

   loyalty_flag  ave_prices   spend_flag  median_days_since_prior_order  
0  New customer    6.367797  low_spender                           20.5  
1  New customer    6.367797  low_spender                           20.5  
2  New customer    6.367797  low_spender                           20.5  
3  New customer    6.367797  low_spender                           20.5  
4  New customer    6.367797  low_spender                           20.5  

[5 rows x 22 columns]

In [66]:
#criteria added
ords_prods_merge2.loc[ords_prods_merge2['median_days_since_prior_order'] > 20, 'freq_flag'] = 'Non_Freq_customer'
ords_prods_merge2.loc[(ords_prods_merge2['median_days_since_prior_order']> 10)& (ords_prods_merge2['median_days_since_prior_order'] <=20 ), 'freq_flag'] = 'Regular_customer'
ords_prods_merge2.loc[ords_prods_merge2['median_days_since_prior_order']<=10, 'freq_flag']= 'Freq_customer'

In [71]:
ords_prods_merge2['freq_flag'].value_counts()

freq_flag
Freq_customer        21559853
Regular_customer      7208564
Non_Freq_customer     3636437
Name: count, dtype: int64

In [72]:
ords_prods_merge=ords_prods_merge2

In [73]:
ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  merge_indicator  \
0                                     Soda  ...             both   
1  Organic Unsweetened Vanilla Almond Milk  ...             both   
2                      Original Beef Jerky  ...             both   
3               Aged White Cheddar Popcorn  ...             both   
4         XL Pick-A-Size Paper Towel Rolls  ...             both   

     price_range_loc    busiest_days busiest_period_of_day max_order  \
0  Mid-range product  Regularly busy        Average orders        10   
1  Mid-range product  Regularly busy        Average orders        10   
2  Low-range product  Regularly busy        Average orders        10   
3  Low-range product  Regularly busy        Average orders        10   
4  Low-range product  Regularly busy        Average orders        10   

   loyalty_flag ave_prices   spend_flag median_days_since_prior_order  \
0  New customer   6.367797  low_spender                          20.5   
1  New customer   6.367797  low_spender                          20.5   
2  New customer   6.367797  low_spender                          20.5   
3  New customer   6.367797  low_spender                          20.5   
4  New customer   6.367797  low_spender                          20.5   

           freq_flag  
0  Non_Freq_customer  
1  Non_Freq_customer  
2  Non_Freq_customer  
3  Non_Freq_customer  
4  Non_Freq_customer  

[5 rows x 23 columns]

In [74]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_4-8.pkl'))

In [1]:
#reassign outliers in 'prices' as missing values

In [ ]:
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan